samza-azure/src/main/java/org/apache/samza/AsyncSystemProducer.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.samza.metrics.Counter;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.eventhub.metrics.SamzaHistogram;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * SystemProducer Helper that can be used to implement SystemProducer that provide async send APIs
+ * e.g. Kinesis, Event Hubs, etc..
+ */
+public abstract class AsyncSystemProducer implements SystemProducer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(NoFlushAsyncSystemProducer.class.getName());
+
+  /**
+   * The constant CONFIG_STREAM_LIST.
+   */
+  public static final String CONFIG_STREAM_LIST = "systems.%s.stream.list";
+  private static final String SEND_ERRORS = "sendErrors";
+  private static final String SEND_CALLBACK_LATENCY = "sendCallbackLatency";
+  /**
+   * The constant AGGREGATE.
+   */
+  protected static final String AGGREGATE = "aggregate";
+
+  /**
+   * The Stream ids.
+   */
+  protected final List<String> streamIds;
+  /**
+   * The Physical to stream ids.
+   */
+  protected final Map<String, String> physicalToStreamIds;
+  /**
+   * The Registry.
+   */
+  protected final MetricsRegistry registry;
+
+  /**
+   * The Pending futures.
+   */
+  protected final Set<CompletableFuture<Void>> pendingFutures = ConcurrentHashMap.newKeySet();
+
+  private final AtomicReference<Throwable> sendExceptionOnCallback = new AtomicReference<>(null);
+
+  private static Counter aggSendErrors = null;
+  private static SamzaHistogram aggSendCallbackLatency = null;
+  private final HashMap<String, SamzaHistogram> sendCallbackLatency = new HashMap<>();
+  private final HashMap<String, Counter> sendErrors = new HashMap<>();
+
+  /**
+   * Instantiates a new Async system producer.
+   *
+   * @param systemName the system name
+   * @param config the config
+   * @param registry the registry
+   */
+  public AsyncSystemProducer(String systemName, Config config, MetricsRegistry registry) {
+    StreamConfig sconfig = new StreamConfig(config);
+    streamIds = config.getList(String.format(CONFIG_STREAM_LIST, systemName));
+    physicalToStreamIds =
+        streamIds.stream().collect(Collectors.toMap(sconfig::getPhysicalName, Function.identity()));
+    this.registry = registry;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public void send(String source, OutgoingMessageEnvelope envelope) {
+    checkCallbackThrowable("Received exception on message send");
+    CompletableFuture<Void> sendResult = sendAsync(source, envelope);
+    pendingFutures.add(sendResult);
+
+    String streamName = envelope.getSystemStream().getStream();
+    String streamId = physicalToStreamIds.getOrDefault(streamName, streamName);
+    long afterSendTimeMs = System.currentTimeMillis();
+
+    // Auto update the metrics and possible throwable when futures are complete.
+    sendResult.handle((aVoid, throwable) -> {
```

Comment:
is this a blocking call?

Reply:
Not sure i understood your questions. Lambda inside sendResult.handle gets called after the send is complete,  It is a callback.  sendResult.handle just hooks up the callback so it is not blocking on the handle logic.